### PR TITLE
Versioning

### PR DESCRIPTION
**Pull request by kryptonite303**
At the moment, the toolchain role doesn't know how to determine what version of the toolchain is currently installed on a host. In order to fix this with least amount of dependencies on the toolchains themselves, child toolchain roles will add a toolchain_version file as part of their tasks.
This PR will allow this role to check that toolchain_version file and decide whether to download a new toolchain or leave it in place by comparing the toolchain_version passed into this role's variables and the toolchain_version file in the toolchain itself.  If there is no version file present, it will download a new toolchain and replace the one in place, which the child toolchain role will add a version file.

Also, in order for this process to work, this role will download the toolchains to the /tmp directory where it can then be moved by the child toolchain roles into the /opt directory since the toolchains have inconsistent naming.

The toolchain roles should now correctly download a different toolchain version if specified.
Additionally, this PR will now always pass the top level directory name to child toolchain roles instead of making it optional and confusing.
Also updated the tests to use a fake java toolchain to speed up tests.

### CODE REVIEW

**Comment by jasonwalsh** [comment]
If I understand this correctly, is this functionality similar to a lock file that serves as a "snapshot" of the currently installed version of a toolchain?

**Comment by kryptonite303** [comment]
It's more like a version file to let the role and us know which version of the toolchain is actually installed. I'm treating the entire toolchain like a piece of software where it can report back to us what version it is. 
I guess a snapshot is another way to look at it?

**Comment by jasonwalsh** [comment]
Yes. It is essentially similar to a `package-lock.json` file created by `npm` or a `Gopkg.lock` file created by `dep`.

**Comment by kryptonite303** [comment]
The only difference there is this isn't recording the versions of the individual packages that makes up the toolchain, only the toolchain as a whole

**Comment by jasonwalsh** [comment]
Sure; though in this case, a toolchain *can* be considered as a package.

**Comment by kryptonite303** [comment]
Actually, there is a "package.lock" sort of file that the old toolchain module uses.
The `/opt/toolchain_metadata/metadata.json` does that. 
It looks like this:
`{"mongo-golang-toolchain/opt": "c2b22e6b09e856aed18bcfddbdccb3aa70c70ece", "mongodbtoolchain/opt/mongodbtoolchain/revisions": "23805b43dd027076b4ae48533ab385e20c61e0cf", "php-toolchain/opt": "37dbfb7b800408897e2dbb6c85ae9e67ce6f7c49", "python-toolchain/opt": "f094c89fae533dd2331cd3b5bf62fc954dfc7d4a", "ruby-toolchain/opt/ruby-{{ revision }}": "a650fe6064a65922754444f3fe04c248cdf50ef1", "perl-toolchain/opt": "ef049c1259f565407eefcf9b76eb8eefb18d2beb", "java-toolchain/opt": "684d8a531d064469d4ca29e831e5929420e486d6"}`
I would run into permissions issues with that file since different distros have different ansible users and it wouldn't always be editable. 
I figured with the version file approach, the roles wouldn't need to depend on an external lock file. The versioning would be more self contained.

File: tasks/main.yml [hunk 1]
```diff
@@ -1,25 +1,37 @@
...
+
+- name: Check version file if toolchain is installed
+  stat:
+    path: "{{ toolchain_final_dest }}/toolchain_version"
+  register: version_file
+  when:
+    - final_dest.stat.exists
+
+- name: Get toolchain version if installed
+  command: cat {{ toolchain_final_dest }}/toolchain_version
```
Comment (zakhark) [inline tasks/main.yml]: Where do we create this file?

Comment (kryptonite303) [reply tasks/main.yml]: This file is created by the child toolchain roles. For example, the golang toolchain role will create `/opt/golang/toolchain_version`

**Pull request merged**